### ISSUE TITLE
[lambda-event-sources] Cross stack reference with dynamodb stream 

### ISSUE THREAD

**Issue by obiwabrakenobi** [aws/aws-cdk#10284]
When I create a dynamodb table in one stack and share that reference to a second stack (same cdk app) and add in the second stack a DynamoEventSource it gives me an error 

```
Stack "s2a-wd-datapoint-importer" cannot consume a cross reference from stack "s2a-wd-plotting". Cross stack references are only supported for stacks deployed to the same environment or between nested stacks and their parent stack

/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/refs.ts:55
    throw new Error(
          ^
Error: Stack "s2a-wd-datapoint-importer" cannot consume a cross reference from stack "s2a-wd-plotting". Cross stack references are only supported for stacks deployed to the same environment or between nested stacks and their parent stack
    at resolveValue (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/refs.ts:55:11)
    at Object.resolveReferences (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/refs.ts:31:24)
    at Object.prepareApp (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/prepare-app.ts:36:5)
    at Object.synthesize (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/synthesis.ts:21:3)
    at App.synth (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/stage.ts:175:23)
    at process.<anonymous> (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/app.ts:112:45)
    at Object.onceWrapper (events.js:422:26)
    at process.emit (events.js:315:20)
    at process.EventEmitter.emit (domain.js:483:12)
    at process.emit (/home/emil/code/watchdog-plotting/node_modules/source-map-support/source-map-support.js:495:21)
```

### Reproduction Steps

Stack1:

```
        this.table= new dynamodb.Table(scope, 'table', {
            partitionKey: {
                name: 'id',
                type: AttributeType.STRING,
            },
            billingMode: BillingMode.PAY_PER_REQUEST,
            tableName: `tabletest`,
            stream: dynamodb.StreamViewType.NEW_IMAGE,
        });
```

Stack2:

```
        lambda.addEventSource(
            new DynamoEventSource(table, {
                startingPosition: StartingPosition.LATEST,
                batchSize: 5,
                maxBatchingWindow: Duration.seconds(10),
                retryAttempts: 1,
            })
        );
```

### What did you expect to happen?

Sharing dynamodb table with streams should work within one cdk app

### What actually happened?

I see the following error. When I remove the EventSource from the lambda everything works fine.

```
Stack "s2a-wd-datapoint-importer" cannot consume a cross reference from stack "s2a-wd-plotting". Cross stack references are only supported for stacks deployed to the same environment or between nested stacks and their parent stack

/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/refs.ts:55
    throw new Error(
          ^
Error: Stack "s2a-wd-datapoint-importer" cannot consume a cross reference from stack "s2a-wd-plotting". Cross stack references are only supported for stacks deployed to the same environment or between nested stacks and their parent stack
    at resolveValue (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/refs.ts:55:11)
    at Object.resolveReferences (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/refs.ts:31:24)
    at Object.prepareApp (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/prepare-app.ts:36:5)
    at Object.synthesize (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/private/synthesis.ts:21:3)
    at App.synth (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/stage.ts:175:23)
    at process.<anonymous> (/home/emil/code/watchdog-plotting/node_modules/@aws-cdk/core/lib/app.ts:112:45)
    at Object.onceWrapper (events.js:422:26)
    at process.emit (events.js:315:20)
    at process.EventEmitter.emit (domain.js:483:12)
    at process.emit (/home/emil/code/watchdog-plotting/node_modules/source-map-support/source-map-support.js:495:21)
```

### Environment

  - **CLI Version      :** 1.62.0 (build 8c2d7fc)
  - **Framework Version:** 1.62.0
  - **Node.js Version:** v12.18.3
  - **OS               :** Ubuntu Server 20.4
  - **Language (Version):** TypeScript (3.9.7)

--- 

This is :bug: Bug Report

**Comment by nija-at** [comment]
The error message states the explanation - `Cross stack references are only supported for stacks deployed to the same environment or between nested stacks and their parent stack`. 

Check your stack configuration and how they're deployed.

Let me know if your stacks are deployed to the same environment and still seeing this error.

**Comment by obiwabrakenobi** [comment]
Hi, its in the same environment. The CDK app creates two stacks and stack1 creates the table and passes them to stack2 where the EventSource is created (like described in the section `Accessing resources in a different stack` in https://docs.aws.amazon.com/cdk/latest/guide/resources.html)

**Comment by nija-at** [comment]
I can't reproduce this issue. 

This CDK app works just fine for me and synthesizes two different stacks -

```ts
import { App, Duration, Stack } from '@aws-cdk/core';
import { Code, Function, Runtime, StartingPosition } from '@aws-cdk/aws-lambda'
import { DynamoEventSource } from '@aws-cdk/aws-lambda-event-sources'
import { AttributeType, BillingMode, StreamViewType, Table } from '@aws-cdk/aws-dynamodb';

const app = new App();

const tablestack = new Stack(app, 'tablestack');

const table = new Table(tablestack, 'table', {
  partitionKey: {
      name: 'id',
      type: AttributeType.STRING,
  },
  billingMode: BillingMode.PAY_PER_REQUEST,
  tableName: `tabletest`,
  stream: StreamViewType.NEW_IMAGE,
});

const functionstack = new Stack(app, 'functionstack');

const fn = new Function(functionstack, 'fn', {
  code: Code.fromInline('foo'),
  runtime: Runtime.NODEJS_12_X,
  handler: 'index.handler',
});

fn.addEventSource(
  new DynamoEventSource(table, {
      startingPosition: StartingPosition.LATEST,
      batchSize: 5,
      maxBatchingWindow: Duration.seconds(10),
      retryAttempts: 1,
  })
);
```

**Comment by obiwabrakenobi** [comment]
lets me check that. If have literally the same setup

**Comment by obiwabrakenobi** [comment]
well not literally the same. As you mentioned the error message was correct but I did not see the problem in the first place.

My setup is like this. I added the env since I could not create a hosted zone

`Cannot retrieve value from context provider hosted-zone since account/region are not specified at the stack level. Either configure "env" with explicit account and region when you define your stack, or use the environment variables "CDK_DEFAULT_ACCOUNT" and "CDK_DEFAULT_REGION" to inherit environment information from the CLI (not recommended for production stacks)`

```
const stack1 = new Stack(app, 'stack1', {
    env: {
        account: process.env.CDK_DEFAULT_ACCOUNT,
        region: process.env.CDK_DEFAULT_REGION,
    },
});

const stack2 = new Stack(
    app,
    'stack2'
);
``` 

But then the environment seems not to be the same for both stacks which makes sense to me. So I added the `env` to stack2 and then it works. 

Thanks for your help! I think it is solved with that
